app/models/spina/admin/conferences/conference.rb

```diff
@@ -127,21 +188,80 @@ def to_event # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
           event = Icalendar::Event.new
           return event if invalid?
 
-          event.dtstart = start_date
-          event.dtstart.ical_param(:value, 'DATE')
-          event.dtend = finish_date
-          event.dtend.ical_param(:value, 'DATE')
+          event.dtstart = Icalendar::Values::Date.new(dates.begin)
+          event.dtend = Icalendar::Values::Date.new(dates.exclude_end? ? dates.end : dates.end + 1.day)
           event.location = location
           event.contact = Spina::Account.first.email
           event.categories = Conference.model_name.human(count: 0)
           event.summary = name
           event
         end
 
-        # @param (see #to_event)
-        # @deprecated Use {#to_event} instead
+        # @return [TZInfo::TimezonePeriod, nil] the time zone period for the conference event
+        def time_zone_period
+          return if start_date.blank?
+
+          start_date.at_beginning_of_day.period
+        end
+
+        # @return [String, nil] the conference as an iCalendar
         def to_ics
-          to_event
+          Rails.cache.fetch [self, presentations, 'calendar'] do
+            calendar = Icalendar::Calendar.new
+            return if invalid?
+
+            calendar.x_wr_calname = name
+            calendar.add_timezone(ical_timezone)
+            eventables_events.each { |event| calendar.add_event(event) }
+            calendar.publish
+            calendar.to_ical
+          end
+        end
+
+        private
+
+        def eventables
+          [self, *events, *presentations]
+        end
+
+        def eventables_events
+          Rails.cache.fetch [self, *presentations, 'events'] do
+            eventables.collect do |eventable|
+              Rails.cache.fetch([eventable, 'event']) { eventable.to_event }
+            end
+          end
+        end
+
+        def ical_timezone
+          Rails.cache.fetch [self, presentations, 'time_zone'] do
+            Icalendar::Timezone.new do |timezone|
+              timezone.tzid = Time.zone.name
```

Comment:
you use Time.zone here, but is the server timezone set at any point? won't this always be UTC? since timezones aren't stored with the dates entered into the db or set in the controller for each request (I don't think?) how are they determined when input in forms in spina?

this might be covered by the tests - I'm afraid I haven't been able to look at this in much detail just yet. And I can definitely understand it working if times are always in GMT/BST (which for our use case they will be).

Reply:
So times are stored in the database in UTC without a timezone, and then Rails converts this to the server's timezone (for the dummy app this is done in config as below (I chose this city because it will never coincide with GMT, and because the difference from GMT is sufficient to realise when times are incorrectly in UTC during testing), but for the main app is done with an environment variable set to London, hence the main app runs in GMT or BST according to daylight savings). https://github.com/jmalcic/spina-admin-conferences/blob/86e1d6bf6ed93b1ecf2dc0056ef81f3ba5f4d798/test/dummy/config/application.rb#L18
The advantage of this implementation is that it allows us to choose (in future) to set a timezone according to the client's preference, user preferences, or some timezone attribute for another record (e.g. a conference could have a timezone attribute and then events and presentations for that conference would be displayed in that timezone). It would be possible to store timezones along with each individual date time, but I can't really imagine a case where that would be useful for a conference—even if the conference was taking place in two institutions straddling two timezones I think you would pick a reference timezone or maybe show times in both timezones (but only those two), or maybe try to get the timezone from the client somehow and use that.

Reply:
Perfect, thanks for the clarification. I agree this is the best way to do it, certainly for now - didn't realise we had a time zone environment var. I saw the use of that city in the dummy when you added it and was curious haha.

Time zone clarification in the front end (i.e. displaying the server reference zone) could potentially be nice (and relatively simple to implement) but that's something for the primer theme gem in any case.